README
METAMOD 2
=========

INTRODUCTION
------------

This directory is the top directory for METAMOD 2. METAMOD is a metadata
catalogue developed by the Norwegian Metrological Institute. METAMOD can
harvest metadata from several different sources and offers a web-based
search interface to the metadata catalogue.

METAMOD is written in Perl with some shell scripting for installation.

MODULES
-------

METAMOD 2.x is divided into seven separate modules:

METAMODBASE     Database for metadata describing scientific datasets and
                web-based management utilities for the database. All software
                for this module are contained in subdirectory: base.

METAMODSEARCH   Web-based search application for searching the database.
                All software for this module are contained in subdirectory:
                search.

METAMODUPLOAD   Web-based data upload application. Includes programs for
                harvesting metadata from netCDF-files. All software for this
                module are contained in subdirectory: upload.

METAMODQUEST    Web-based questionnaire for entering metadata into the
                datatbase. All software for this module are contained in
                subdirectory: quest.

METAMODPMH      OAI-PMH server for metadata harvesting

METAMODHARVEST  OAI-PMH client for metadata harvesting

METAMODTHREDDS  Automatic generation and maintenance of a thredds catalog XML
                file that makes new datasets visible through a thredds server.
                Note that use of a thredds server as an interface to the data
                repository is not dependent on this module, but if thredds is
                used without this module, a thredds catalog has to be
                maintained manually.

The modules can be used separately and in combination. An actual usage of
the software must be the result of an installation process, where the
modules to be used are chosen. An instance of the installed software,
comprising one or more modules, is called an application. An application not
using the METAMODBASE module must identify another application that use the
METAMODBASE module. Then the application will use the database in this other
module. Several applications may, in this way, share a common database, for
which one of the applications has the main responsibility.

DEPENDENCIES
------------

This software depends on the following external software which is not part
of the METAMOD 2.x distribution:

- PostgreSQL database system. Tested on version 8.4

- Apache 2.x web server with mod_proxy installed

- libxml and libxslt library from xmlsoft.org

- Proj4 library (http://proj.maptools.org)

- NetCDF software 3.x (x >= 5): libnetcdf, ncdump and ncgen (METAMODUPLOAD)

- Perl v5.8 with a whole host of Perl modules, of which Moose, Catalyst,
  Template Toolkit, DBIx::Class, HTTP::OAI, TheSchwartz and Data::FormValidator
  are the most important. A complete list should be found in catalyst/Makefile.PL.

METAMOD 2.x has been developed and tested using Ubuntu Linux 10.04, but
should work on most operating systems supporting the mentioned libraries.


SYSTEM OVERVIEW (OBSOLETE - FIXME)
---------------

For each module there is a more detailed README file describing the
operations of that module. This paragraph gives a summary of the operations
of a full application comprising all modules.

An application must be configured and installed as described later in this
README-file. After installation, if the application is a fresh one (where
no database for the application already exists), a new database must be
initialized and filled with static data. Also, if not already done for the
PostgreSQL installation, two database users must be defined. The default names
of these users are 'admin' and 'webuser', but these names are configurable.
The users can be created by the createusers.sh script, and this must be done
prior to creating the database.

Initializing the database and loading the static data are done by running the
init/create_and_load_all.sh script as described in base/README. The static
data loaded into the database are found in an XML file: staticdata/searchdata.xml
and defines the search categories by which it is possible to search the database.
This file must be found in the application directory of the application owning
the database. All applications using this database will be confined to the
search categories defined by this file. The example application included in
the source tree has one version of this file that is based on the
application installed for the DAMOCLES project.

The software also need a runtime environment that must be set up by the
prepare_runtime_env.sh shell script. This script creates a directory (the
webrun directory) if not already found. Several subdirectories in the webrun
directory are also created. This shell script also copies a .htaccess file
into the htdocs/adm directory to protect the administration utilities from
unauthorised usage. If the prepare_runtime_env.sh script has already been run
during an earlier installation process, it is not neccessary to do it again
(unless a new version of the software requires changes in the runtime
environment). On the other hand, a re-run of this script will do no harm.

After initialization, loading of static data and setting up the runtime
environment, the web interfaces for the METAMODSEARCH, METAMODUPLOAD and
METAMODQUEST modules should be up and running.

To fill the database with metadata, any of the METAMODUPLOAD, METAMODQUEST
and/or METAMODHARVEST modules must be applied. These modules produce XML files
with metadata describing datasets (see base/README_XML).

Datasets are organized hierarchically in two levels. The top level contains
directory datasets. Each directory dataset may own one or more file datasets
at the second (bottom) level. File datasets will correspond to physical files
found in a local or external data repository. Directory datasets may correspond
to directories in the data repositories where these files reside. The database
will also contain independent datasets that own no file dataset on the second
level. These datasets belong to the top level, and are considered directory
datasets even if they correspond to no physical directory.

The metadata comprising a dataset are found both in the database and in a set
of XML files in the local file system. Two XML files exist for each
dataset. With the METAMODQUEST routine, all the content of the two XML files has
to be manually entered into an HTML form. The METAMODUPLOAD routine creates the
XML files based on netCDF files uploaded to a data repository. Uploads may either
be done interactively through the METAMODUPLOAD web interface, or by FTP
(usually from a script run by the data provider). Each netCDF file will
comprise one file dataset. The data provider decides how files and file datasets
are organized into directories and directory datasets. The METAMODHARVEST module
periodically harvest metadata from external sources. The harvested metadata are
saved as XML files.

The collection of XML files comprise at any time, a full backup of the database.
A script exists that will recreate the database from the XML files.

Two perl scripts are responsible for extracting the metadata from the netCDF
files and load them into the database. The first script, upload_monitor.pl,
belongs to the METAMODUPLOAD module. This script extracts the metadata from
a batch of netCDF files and creates/updates the corresponding XML files. The
other script, import_dataset.pl, belongs to the METAMODBASE module. This
script monitors a number of directories to see if XML files are created or
updated, and loads the XML files into the database. Both scripts are running
in the background. They wake up at regular intervals to see if any work
are pending. These perl scripts have to be manually started by the
'metamodInit.sh start' shell script. They will run until another shell script,
'metamodInit.sh stop', stops the scripts in an orderly fashion.

DIRECTORY STRUCTURE (OBSOLETE - FIXME)
-------------------

The top METAMOD 2.x directory (the directory where this README file resides)
is divided into one subdirectory for each module (base, search, upload, quest,
pmh, harvest and thredds). There is also a subdirectory (common) where files
used by many modules resides, and a subdirectory (test) containing a test suite
for periodically excercising the software while software developement are going
on. In addition, one subdirectory (app) contains an example application.

The module directories shares a common structure. During installation of an
application, a new directory is created (the target directory), and all module
directories are merged into this target directory. In each module directory
there is a htdocs subdirectory. The htdocs directories for all the selected
modules are in the same way merged into one htdocs subdirectory in the target
directory.  To avoid name collisions, the actual HTML and PHP files etc. are
contained within subdirectories of the htdocs directory. For example, all such
files are contained in the directory htdocs/sch for the METAMODSEARCH module,
and htdocs/upl for the METAMODUPLOAD module. On the target directory, there
will be a htdocs directory containing both the sch and the upl subdirectories.
Some htdocs subdirectories may be shared among several modules. For example,
the METAMODBASE module and the METAMODQUEST module both have files in the adm
subdirectory. Special care must be taken for these files to avoid name
collisions.

Each module directory must also contain a file 'filelist.txt' which
lists all files comprising the module.

Each METAMOD application must be configured in a separate directory tree.
This application directory tree will generally be independent of the METAMOD
source tree (where this README file resides at the top level). But an example
of a METAMOD application directory (app/example) is included in the source
tree as a template for real application directories.

In each application directory there must be a configuration file called
'master_config.txt', and a 'filelist.txt' file containing a list of files that
are specific for the application (image files, style sheets etc.). An
application directory has the same structure as any of the module directories,
and contains a htdocs subdirectory divided into sch, upl etc. as needed.
Any file found in the application directory (and included in the 'filelist.txt'
file) will replace a corresponding file in a module directory. This makes it
possible to taylor a METAMOD application to any specific need, but usually this
mechanism is only used for presentation purposes (stylesheets etc.) and for
static data used to setting up the search interface.

CONFIGURATION (OBSOLETE - FIXME)
-------------

PHP CONFIGURATION (OBSOLETE - FIXME)
-----------------
The following values should be configured in the php.ini file

track_errors = On
upload_max_filesize = 100M
post_max_size = 100M

METAMOD CONFIGURATION (OBSOLETE - FIXME)
---------------------

The source files within the METAMOD 2.x directories contain a variety of file
types (PHP, Perl, XML, HTML/XHTML, CSS and shell-script files, some of
them with embedded SQL). The source files are not intended to be used
directly, but must be copied to a target directory tree before the application
can run. Some of the files will be modified (see below) before ending up in the
target directory. The reason for this is to apply a unified configuration method
for all the different file types.

The following file is the master configuration file for one application:

    master_config.txt

This file contains lines as follows:

VARNAME = VALUE

'VARNAME' must be all uppercase letters [A-Z], underscore or digits. Any number
of white space characters may separate 'VARNAME' and '=', and also '=' and
'VALUE'. 'VALUE' starts with the first non-whitespace character and ends with
the last non-whitespace character on the line. Additional lines may be
appended to 'VALUE' immideately after the initial 'VARNAME = ...' line. Such
lines must start with a space character. No whitespace are removed from such
appended lines.

VALUEs may be empty, which sometimes indicate that some aspect of the software
is turned off. An empty value is indicated by having no characters (or only
whitespace characters) after the '=' character.

In addition, comment lines, starting with '#', may be found. Blank lines are
ignored.

In the source files, all text strings like:

[==VARNAME==]

will be substituted with the corresponding value when the files are modified and
moved to the target directory. Such text strings may also be found within 'VALUE'
strings, and they will be substituted as appropriate. When a source file contains
a [==VARNAME==] string, but no corresponding VARNAME definition is found in the
configuration file, the [==VARNAME==] string is substituted with an empty string,
and a warning is given during installation.

Two variables are mandatory in master_config.txt:

TARGET_DIRECTORY      This is the directory name of the target directory, the
                      top level of the target directory tree.
SOURCE_DIRECTORY      This is the directory name of the source directory where
                      this README file resides.

The following variables are used to select the modules that will comprise the
application. They also identify the absolute path of the source directory for
each module (i.e. they will be subdirectories of the SOURCE_DIRECTORY):

METAMODBASE_DIRECTORY
METAMODSEARCH_DIRECTORY
METAMODUPLOAD_DIRECTORY
METAMODQUEST_DIRECTORY
METAMODPMH_DIRECTORY
METAMODHARVEST_DIRECTORY
METAMODTHREDDS_DIRECTORY

If any of these variables are missing, the corresponding module will not be part
of the application.

Each application delivering data to a database (an application with the
METAMODUPLOAD, METAMODHARVEST and/or the METAMODQUEST module) must have a short
acronym that uniquely identify the application for the database. This acronym is
included in the master_config.txt file as the value of variable APPLICATION_ID.

INSTALLATION (OBSOLETE - FIXME)
------------

To install an application, the following installation script (located in the top

CHANGES

============================
  METAMOD CHANGELOG
============================

2.14   Thu, 22 Oct 2015
  * ported to Ubuntu 14.04 (Trusty Tahr) and PostgreSQL 9.3
  * automatic detection of PSQL version during installation

2.13   Mon Jun 3 2013
  * dynamic search map using WMS
  * improved support for background maps in different projections
  * automatic building and installation of Perl dependencies using Carton
  * support for Ubuntu 12.04 Precise Pangolin and PostgreSQL 9.1
  * fix for old GetLegendGraphics bug to work with new version of THREDDS
  * upgraded to latest Catalyst version (5.9)
  * support for boolean operators, phrases and wildcards in text search
  * easier configuration (lsconf, virtualenv, sanity checker)
  * support for manual configuration and serving static files in Apache (again)
  * WMS search map thumbnails
  * support for custom validators in OAI-PMH
  * support for two-dimensional arrays in timeseries plots and ascii/json download
  * WMS slideshow, frame stepping controls
  * template texts fully customizable per site
  * download external files, generates client file download scripts in various languages
  * support for nGinX and CheckMK

2.12   Mon Jan 14 2013
  * projection selector and timeslider in wms client
  * visualization of non-gridded timeseries
  * fimex support for reading via opendap
  * subsetting and (experimental) reprojection via OPeNDAP
  * bugfix for wgs84 map in basket

2.11    Thu Aug 31 2012
    Improved WMS client with many new features:
  * new advanced layer selector with separate legends and time axis
  * multiple transparent daya layers can be displayed simultaneously
  * layers from different datasets and sources can be combined in client using collection basket
  * timeslider allows for animation of multiple layers simultaneously
  * support for many more projections (depending on data source WMS server)
  * added support for WMS 1.1.1 (in addition to 1.3.0)

2.10    Tue 24 Jan 2012
  * Simplified installation - now running directly from source, Debian installer
  * WMS visualization of time series, aggregated datasets, collection basket
  * Final PHP module (OAI-PMH) rewritten in Perl
  * Added config defaults, simplifying site administration

2.9     Thu Sep 8 09:33:55 2011
  * Allowing near-realtime web upload
  * Queue system implemented for background jobs
  * Cron scheduling for Harvester, FTP upload
  * Improvements in the metadata backend
  * Each site must now run separate userbase

2.8     Wed May 18 10:50:41 2011
  * Stable release using Catalyst/Perl web frontend
  * Hashed passwords for improved security

2.7     Internal milestone, not released
  * Technology preview using Catalyst/Perl for web interface instead of PHP

2.6     Thu Nov 25 08:00:00 2010
  * Connecting to SRU2JDBC for ISO23950 support
  * Dataset subscription via e-mail and RSS support
  * Improved ISO19115/19139 support compatible with WIS and GeoNetworks
  * WMS improvements
  * User database added

2.5     Tue May 11 11:15:51 2010
  * WMS presentation of data
  * WMS preview in client
  * Registration of files in existing repository
  * Search in different world maps

2.4     Fri Mar 12 13:11:59 2010
  * Internal release number only

2.3     Wed Nov 18 09:57:57 2009
  * general text search
  * methods for re-projecting 2D data (fimex)
  * methods for connecting to WMS server
  * module for automatic update of THREDDS catalog.
  * improvements in PMH server and dataset administration.

2.2     Wed Apr 1 13:54:53 2009
  * Implementation of a two-level view of datasets. At the highest level,
    datasets representing several files are described. Details of each file
    are described at the second level

2.1     Fri Dec 19 14:31:13 2008

2.0     Mon Jul 21 07:44:07 2008

1.x     2002-2007
    During 2001/2002, the Norwegian Meteorological Institute (METNO) implemented
    a web based tool (MetaMod) for searching data archived under the NoSerC
    project (http://noserc.met.no/). This tool was adapted for the DAMOCLES
    project in 2006/2007. The web interface and the overall functionality of the
    tool were very much improved during the DAMOCLES development, and just a
    small part of the original code base has remained unchanged.


update_target.pl
#!/usr/bin/perl -w

=begin LICENCE

----------------------------------------------------------------------------
  METAMOD - Web portal for metadata search and upload

  Copyright (C) 2008 met.no

  Contact information:
  Norwegian Meteorological Institute
  Box 43 Blindern
  0313 OSLO
  NORWAY
  email: egil.storen@met.no

  This file is part of METAMOD

  METAMOD is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  METAMOD is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with METAMOD; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
----------------------------------------------------------------------------

=end LICENCE

=cut

use strict;
use File::Copy;
use File::Path;
use File::Spec;
use FindBin qw($Bin);
use Cwd;

use lib "$Bin/common/lib";

#
#  Check command line arguments (FIXME: rewrite using Getopt::Long)
#
my %commandline_options = (); # Hash containing commandline options of the form --opt or --opt=val
                              # $commandline_options{"--opt"} == "val" (or empty)
my $appdir;
my $argcount = 0; # Counts number of normal arguments (not containing '=')
my $optcount = 0; # Counts number of option arguments (containing '=')
foreach my $cmdarg (@ARGV) {
   if ($cmdarg =~ /^([^=]+)=(.*)$/) {
      my $opt = $1;
      my $val = $2;
      $commandline_options{$opt} = $val;
      $optcount++;
   } else {
      $appdir = $cmdarg;
      $argcount++;
   }
}
if ($argcount != 1 || ($optcount > 0 && $optcount != 2) ||
      ($optcount > 0 && !exists($commandline_options{'--from'})) ||
      ($optcount > 0 && !exists($commandline_options{'--to'}))) {
   die "\nUsage:\n\n1.     $0 application_directory\n\n" .
       "where 'application_directory' is the name of a directory containing the application\n" .
       "specific files. Inside this directory, there must be a master_config.txt file.\n\n" .
       "2.     $0 --from=sourcefile --to=targetfile application_directory\n\n" .
       "If these two options are present, the normal behaviour is suspended. The only thing\n" .
       "that happens are copying with substitutions (according to the master_config.txt file)" .
       "from the sourcefile to the targetfile.\n\n";
}
#
#   Read the configuration file:
#
my $configMaster = 'master_config.txt';
my $configfile = "$appdir/$configMaster";
my $appfilelist = $appdir . '/filelist.txt';
my $appName = (File::Spec->splitdir( $appdir))[-1];
my $version;
my $majorVersion;
{
    open my $verFH, "$Bin/VERSION" or die "cannot read VERSION-file at $Bin/VERSION: $!\n";
    while (defined (my $line = <$verFH>)) {
        if ($line =~ /version\s+(\d+\.\w+)\.(\w+)\s+/i) {
            $version = $1 . '.' . $2;
            $majorVersion = $1;
            last;
        }
    }
    unless (defined $majorVersion) {
        die "could not read version from VERSION file\n";
    }
}
if ($optcount == 0) { # Supress this message if only "copy with substitution" is done
   print "\nConfiguring application '$appName' with Metamod $majorVersion ($version)\n\n";
}

#######################
#
# Process master_config
#
my %conf;
$conf{MAJORVERSION} = $majorVersion;
my $missing_variables = 0;

#  Open config file for reading
my $config_modified = (stat($configfile))[9];
unless (-r $configfile) {die "Can not read from file: $configfile\n";}
open (CONFIG,$configfile);

my %newfilenames = ();
my $value = "";
my $varname = "";
my $origname = "";
my $newname = "";
my $line = "";

#  Loop through all lines in config file:
while (<CONFIG>) {
   chomp($_);
   $line = $_;

   # check for something (looks like we're closing a multi-line assignment?)
   if ($line =~ /^[A-Z0-9_#!]/ && $varname ne "") {
      if (length($origname) > 0) {
         $conf{$varname . ':' . $origname . ':' . $newname} = $value;
         $newfilenames{$origname . ':' . $newname} = 1;
      } else {
         $conf{$varname} = $value;
      }
      $varname = "";
   }

   if ($line =~ /^([A-Z0-9_]+)\s*=(.*)$/) {  # this looks like a single-line assignment
      $varname = $1; # First matching ()-expression
      $value = $2; # Second matching ()-expression
      $value =~ s/^\s*//;
      $value =~ s/\s*$//;
   } elsif ($line =~ /^!substitute_to_file_with_new_name\s+(\S+)\s+=>\s+(\S+)\s*$/) {
      $origname = $1;
      $newname = $2;
   } elsif ($line =~ /^!end_substitute_to_file_with_new_name\s*$/) {
      $origname = "";
      $newname = "";
   } elsif ($line !~ /^#/ && $line !~ /^\s*$/) {  # append value in multi-line assignment
      $value .= "\n" . $line;
   }
}

if ($varname ne "") {  # fall-off multi-line assignment (end of file)
   $conf{$varname} = $value;
}

close (CONFIG);
# finished processing config

#
# Start processing output files
#

if ($optcount > 0) {
   # Only copy (with substitutions) one file to a targetfile. Then exit:
   my $sourcefile = $commandline_options{'--from'};
   my $targetfile = $commandline_options{'--to'};
   print "Copy and substitute from $sourcefile to $targetfile\n";
   &substcopy($sourcefile,$targetfile);
   exit;
}

# now we need to copy a bunch of files into target dir
if (!exists($conf{'TARGET_DIRECTORY'})) {
   die "TARGET_DIRECTORY is not defined in $configfile";
}

#
#  Create an array of pathes (@flistpathes). Each element in this array
#  is the name of a file containing a list of source files to be
#  copied to the target directory.
#
my @flistpathes = ();
my $targetdir = $conf{'TARGET_DIRECTORY'};
$targetdir = &substituteval($targetdir);
die "Cannot write to target dir $targetdir" unless -w $targetdir;

#
# Install the config-file
#
if (! -f "$targetdir/$configMaster" or ($config_modified > (stat _)[9])) {
   print "Copy to $targetdir/$configMaster\n";
   copy($configfile, "$targetdir/$configMaster") or
      die "Could not copy $configfile to $targetdir: $!";
}

# install link from /etc/metamod-$majorVersoin/$appName.cfg to $targetdir/$configMaster
# NOTE: this only works when running as root, and should not be run on development servers
#{
#    my $etcDir = "/etc/metamod-$majorVersion";
#    if (-d $etcDir) {
#        my $etcConfig = "$etcDir/$appName.cfg";
#        unlink $etcConfig; # need root for this
#        symlink("$targetdir/$configMaster", $etcConfig)
#            or print STDERR "unable to create symlink from $etcConfig -> $targetdir/$configMaster\n";
#    } else {
#        print STDERR "Cannot create configuration-symlink in $etcDir: no such directory\n\n";
#    }
#}

#
# get list over files to process
#
if (!exists($conf{'SOURCE_DIRECTORY'})) {
   die "SOURCE_DIRECTORY is not defined in $configfile";
}
my $sourcedir = $conf{'SOURCE_DIRECTORY'};
$sourcedir = &substituteval($sourcedir);
if (-r $sourcedir . '/common/filelist.txt') {
   push (@flistpathes,$sourcedir . '/common/filelist.txt');
}
foreach my $module qw(METAMODBASE METAMODUPLOAD METAMODQUEST METAMODPMH METAMODHARVEST METAMODTHREDDS) {
   if (exists($conf{$module . '_DIRECTORY'}) and $conf{$module . '_DIRECTORY'} ne "") {
      my $moduledir = $conf{$module . '_DIRECTORY'};
      $moduledir = &substituteval($moduledir);
      push (@flistpathes,$moduledir . '/filelist.txt');
   }
}
push (@flistpathes, $appdir . '/filelist.txt');

#
# install Catalyst files before $appdir to avoid clobbering custom files
#
&install_catalyst;

#
# start processing source files
#
my %copied_targetnames = ();
foreach my $filelistpath (@flistpathes) {
#
#  Read the list of file names to be copied.
#  For each of these files where the modification time is larger
#  than the corresponding file in the target directory tree,
#  the file is copied to the target directory tree. While copying,
#  substitutions are made according to the configuration file.
#
   my $topdir = $filelistpath;
   $topdir =~ s:/[^/]*$::; # $topdir is the directory containing the
                           # current filelist.txt file
   # Open source file for reading
   unless (-r $filelistpath) {die "Can not read from file: $filelistpath\n";}
   open (FILES,$filelistpath);
   print "--- Processing $filelistpath:\n";

   # Loop through all lines in source file:
   while (<FILES>) {
      chomp($_);
      next unless /\w+/;
      my $filename = $_;
      my $ch1 = substr($filename,0,1);
      next if $ch1 eq '#'; # allow comments in file
      if ($ch1 eq '=' ) { # File will be copied unmodified
         $filename = substr($filename,1);
      }
      my $srcfilepath = $topdir . '/' . $filename;
      if (!( -e $srcfilepath)) {
         print "Unknown file in $filelistpath: $filename [$srcfilepath]\n";
      } else {
         my @targetfilenames = ();
         my $rex = '^' . $filename . ':';
         foreach my $oldandnew (grep {/$rex/} keys %newfilenames) {
            my ($old,$new) = split(/:/,$oldandnew);
            push @targetfilenames, $new;
         }
         if (scalar @targetfilenames == 0) {
            push @targetfilenames, $filename;
         }
         foreach my $targetname (@targetfilenames) {
            my $targetfile = $targetdir . '/' . $targetname;
            if ( exists($copied_targetnames{$targetname})
                 || !( -e $targetfile)
                 || (stat($srcfilepath))[9] > (stat($targetfile))[9]
                 || ($ch1 ne '=' and $config_modified > (stat($targetfile))[9])
               ) {
               # File $filename or the config file is modified later than $targetfile:
               # (or the file has already been copied from another filelist - this
               # ensures that files from the lattermost part of the sequence of filelists
               # are chosen).
               my $dirpath = $targetdir . '/' . $targetname;
               $dirpath =~ s/\/[^\/]*$//;
               mkpath($dirpath);
               #my $current_count = $missing_variables;
               if ($ch1 eq '=') { # copy contents verbatim
                  print "Copy to $targetdir/$targetname\n";
                  copy($srcfilepath, $targetdir . '/' . $targetname) or
                           die "Could not copy $srcfilepath to $targetdir: $!";
               } else { # process contents before copying
                  print "Copy and substitute to $targetdir/$targetname\n";
                  &substcopy($topdir, $filename, $targetdir, $targetname);
               }
               if ($targetfile =~ /([^.]+)$/) {
                  if ($1 eq "pl" || $1 eq "sh") {
                     chmod 0755, $targetfile;
                  }
                  #if ($1 eq "sh" and $current_count < $missing_variables) {
                  #   die "Fatal error: Missing config definitions for shell script $filename";
                  #} # not really a good idea since you get errors for scripts you never will run (e.g. harvester and thredds)
               }
               $copied_targetnames{$targetname} = 1;
            }
         }
      }
   }
   close (FILES);
}


if ($missing_variables > 0) {
   print "NOTE: All [==...==] constructs found that were not defined in the configuration file\n" .
                "      were substituted with empty values\n";
}

my $appuser = &substituteval($conf{'APPLICATION_USER'}) if exists $conf{APPLICATION_USER};
my $webrun = &substituteval($conf{'WEBRUN_DIRECTORY'}) if exists $conf{WEBRUN_DIRECTORY};
if ($appuser) {
   my $owner = getpwuid( (stat($webrun))[4] );
#   print STDERR "webrun dir '$webrun' is owned by user '$owner'\n";
   warn "Webrun directory ($webrun) should be owned by user $appuser (not $owner)" unless $owner eq $appuser;
}

system "$targetdir/scripts/gen_httpd_conf.pl", $targetdir;

#----------------------------------------------------------------------
# END
#----------------------------------------------------------------------

sub install_catalyst {
   # install the catalyst application
   my $cata_dir = "$sourcedir/catalyst";
   my $cata_inst_dir = "$targetdir";
   my $catalyst_lib_dir = "$cata_inst_dir/lib";

   my @required_dirs = ("$cata_inst_dir/scripts", "$cata_inst_dir/lib", "$cata_inst_dir/root");

   foreach my $dir (@required_dirs){
      if( ! -e $dir ){
         mkdir $dir or die "Failed to create directory: $dir";
      }
   }

   dircopy("$cata_dir/script/*", "$cata_inst_dir/scripts");
   dircopy("$cata_dir/lib/*", "$cata_inst_dir/lib");
   dircopy("$cata_dir/root/*", "$cata_inst_dir/root");

   open my $DUMMY_MAKE, '>', "$cata_inst_dir/Makefile.PL";
   print $DUMMY_MAKE "#This is a dummy make file so that catalyst finds the root directory as in source";
   close $DUMMY_MAKE;

}

sub substcopy {
   # copies a file with inline processing
   my $inputdir;
   my $inputfile = "";
   my $inputpath;
   my $outputdir;
   my $outputfile = "";
   my $outputpath;

   # Check number of arguments
   if (scalar @_ == 4) {
      $inputdir = $_[0];
      $inputfile = $_[1];
      $outputdir = $_[2];
      $outputfile = $_[3];
      $inputpath = "$inputdir/$inputfile";
      $outputpath = "$outputdir/$outputfile";
   } elsif (scalar @_ == 2) {
      $inputpath = $_[0];
      $outputpath = $_[1];
   } else {
      die "\nsubstcopy:     Wrong number of arguments\n\n";
   }

   # Open target file for writing
   open (OUT,">$outputpath") or die "Couldn't open $outputpath for writing";

   # Open source file for reading
   unless (-r "$inputpath")
          {die "Can not read from file: $inputpath\n";}
   open (IN,"$inputpath");
   while (<IN>) {
      chomp($_);
      my $sline = $_;
      $sline = &substituteval($sline,$inputfile,$outputfile);
      print OUT $sline . "\n";
   }
   close (IN);
   close (OUT);
   # copy executable flag from source (svn)
   chmod(0755, $outputpath) if -x $inputpath;
}

sub substituteval {
   # process subtitution, either on string or source file
   my $textline;
   my $ifil = "";
   my $ofil = "";
   if (scalar @_ == 3) {
      ($textline,$ifil,$ofil) = @_;
   } else {
      ($textline) = @_;
   }
   my $scount = 0;
   my $substituted;
   my $vname = "";
   do {
      my $valuefound = 0;
      $substituted = 0;
      if ($textline =~ /\[==([A-Z0-9_]+)==\]/) {
         $vname = $1; # First matching ()-expression
         my $value = "";

         #  Check if substitution key exists in config hash
         if (length($ifil) > 0 &&
                  exists($conf{$vname . ':' . $ifil . ':' . $ofil})) {
            $value = $conf{$vname . ':' . $ifil . ':' . $ofil};
            $valuefound = 1;
         } elsif (exists($conf{$vname})) {
            $value = $conf{$vname};
            $valuefound = 1;
         }
         my $reg = "\\[==" . $vname . "==\\]";
         if ($valuefound) {
#
#           Substitute all occurences of a match:
#
            $textline =~ s/$reg/$value/mg;
         } else {
            if (length($ifil) > 0) {
               print "WARNING: [==" . $vname . "==] in $ifil not configured\n";
            } else {
               print "WARNING: [==" . $vname . "==] not configured:\n";
               print "         $textline\n";
            }
            $textline =~ s/$reg//mg;
            $missing_variables++;
         }
         $substituted = 1;
         $scount++;
      }
   } while ($substituted == 1 && $scount < 20);
   if ($scount >= 20) {
      die "ERROR: Circular substitutions in $configfile involving [==$vname==]\n";
   }
   return $textline;
}

=head2 dircopy($src, $dest)

Directory copy implemented via Unix 'rsync' command.

IMPLEMENTATION NOTE: This could be implemented in a more portal fashion with
C<File::Copy::Recursive>, but we don't want any none core dependencies in
update_target.pl.

=cut
sub dircopy {
    my ($src, $dest) = @_;

    my $output = qx/rsync -aC $src $dest/;
    print "$output\n" if $output;

}

#----------------------------------------------------------------------

=head1 NAME

B<update_target.pl> - Metamod installer/tester

=head1 VERSION

[% VERSION %], last modified $Date: 2008-10-06 11:28:08 $

=head1 DESCRIPTION

This script copies the files listed in the various filelist.txt manifests into the
corresponding dirs in the target.
Also performs macro expansion of several shell scripts and other files, substituting
variables from master_config.txt.

=head1 USAGE

 trunk/update_target.pl application_directory
 trunk/update_target.pl --from=sourcefile --to=targetfile application_directory

=head1 OPTIONS

=head2 Parameters

=over 4

=item application_directory

'application_directory' is the name of a directory containing the application
specific files. Inside this directory, there must be a master_config.txt file.

=item --from

=item --to

If these two options are present, the normal behaviour is suspended. The only thing
that happens are copying with substitutions (according to the master_config.txt file)
from the sourcefile to the targetfile.

=back

=head1 LICENSE

Copyright (C) 2010 The Norwegian Meteorological Institute.

METAMOD is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

=cut
